contrib/diurnal/dc.go

```diff
@@ -0,0 +1,293 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// An external diurnal controller for kubernetes. With this, it's possible to manage
+// known replica counts that vary throughout the day.
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+
+	"github.com/golang/glog"
+)
+
+type timeCount struct {
+	time  time.Duration
+	count int
+}
+
+func (tc timeCount) String() string {
+	return fmt.Sprintf("(%s, %d)", tc.time, tc.count)
+}
+
+func parseTimeCounts(times string, counts string) ([]timeCount, error) {
+	ts := strings.Split(times, ",")
+	cs := strings.Split(counts, ",")
+	if len(ts) != len(cs) {
+		return nil, fmt.Errorf("provided %d times but %d replica counts", len(ts), len(cs))
+	}
+	var tc []timeCount
+	prev := time.Duration(-1)
+	for i := range ts {
+		t, err := time.ParseDuration(ts[i])
+		if err != nil {
+			return nil, err
+		}
+		if t < 0 {
+			return nil, errors.New("times must be non-negative")
+		}
+		if t <= prev {
+			return nil, errors.New("times must be in increasing order")
+		}
+		prev = t
+		c, err := strconv.ParseInt(cs[i], 10, 64)
+		if c < 0 {
+			return nil, errors.New("counts must be non-negative")
+		}
+		if err != nil {
+			return nil, err
+		}
+		tc = append(tc, timeCount{t, int(c)})
+	}
+	return tc, nil
+}
+
+type Scaler struct {
+	timeCounts []timeCount
+	ctlr       *api.ReplicationController
+	selector   labels.Selector
+	start      time.Time
+	pos        int
+	done       chan struct{}
+}
+
+func NewScaler(tc []timeCount, selector labels.Selector) (*Scaler, error) {
+	smoothed, err := smoothChanges(tc, *maxScaleRate)
+	if err != nil {
+		return nil, err
+	}
+	return &Scaler{timeCounts: smoothed, selector: selector}, nil
+}
+
+// smoothChanges will add intermediate changes so that the rate of change will not
```

Comment:
Removing this in the next version. Will do simple jumping between replica counts.